### ISSUE TITLE
Unable to freeze a mocked delegate with AutoMoq

### ISSUE THREAD

**Issue by ecampidoglio** [AutoFixture/AutoFixture#742]
AutoMoq has been able to create mocks for delegate types for a while now (see f98ae99).

However, _freezing_ a mocked delegate doesn't seem to work. 😢

Consider this failing test:
 
```csharp
[Fact]
public void FixtureCanFreezeMockOfFunc()
{
    var fixture = new Fixture().Customize(new AutoMoqCustomization());
    var frozen = fixture.Freeze<Mock<Func<string>>>();

    var actual = fixture.Create<Func<string>>();
    
    Assert.Same(frozen.Object, actual); ✖️
}
```

One would expect that a request for `Func<string>` would be satisfied by the frozen `Mock<Func<string>>` but, unfortunately, that's not the case.

## Background

Right now, the `AutoMoqCustomization` configures AutoFixture to provide instances of mocks in response to requests for [abstract classes and interfaces](https://github.com/AutoFixture/AutoFixture/blob/93d339bb63ef7f00d13d9373c4ce59f9419fc76b/Src/AutoMoq/MockRelay.cs#L115-L116) through the [`MockRelay`](https://github.com/AutoFixture/AutoFixture/blob/93d339bb63ef7f00d13d9373c4ce59f9419fc76b/Src/AutoMoq/MockRelay.cs):

```csharp
// MockRelay.cs
private class IsMockableSpecification : IRequestSpecification
{
    public bool IsSatisfiedBy(object request)
    {
        var t = request as Type;

        if (t == null)
        {
            return false;
        }

        return t.IsAbstract || t.IsInterface;
    }
}
```

Given that, the first solution that came to mind was to add delegates to that expression using the internal [`IsDelegate()`](https://github.com/AutoFixture/AutoFixture/blob/ab829640ed8e02776e4f4730d0e72ab3cc382339/Src/AutoMoq/MockType.cs#L46-L54) extension method:

```csharp
return t.IsAbstract || t.IsInterface || t.IsDelegate();
```

However, that alone didn't solve the problem. The reason is that the `AutoMoqCustomization` appends the `MockRelay` to the list of _residue collectors_; as the name implies, that list contains the specimen builders to be invoked _after_ all other builders in the pipeline have failed to satisfy the request. This makes sense because we don't want AutoFixture to provide mocks for interface types that _are_ supported, like for example `IEnumerable<T>`.

In this particular situation, a request for `Func<string>` is handled by the [`DelegateGenerator`](https://github.com/AutoFixture/AutoFixture/blob/ab829640ed8e02776e4f4730d0e72ab3cc382339/Src/AutoFixture/Kernel/DelegateGenerator.cs)—one of the built-in specimen builders—and therefore never reaches the `MockRelay`.

## Workaround

Since moving `MockRelay` to the top of the pipeline in the `AutoMoqCustomization` isn't an option (it would break the existing behavior) the only workaround is to do it in a _custom_ customization and use that in scenarios where you need to freeze a mocked delegate type.

This new `EnableAutoMoqForDelegatesCustomization` would have to do two things:

1. Create a `MockRelay` that includes requests for delegate types.
2. Insert that `MockRelay` at the beginning of the pipeline, so that it takes precedence over the built-in `DelegateGenerator`.

Here's a sample implementation:

```csharp
public class EnableAutoMoqForDelegatesCustomization : ICustomization
{
    public void Customize(IFixture fixture)
    {
        if (fixture == null)
        {
            throw new ArgumentNullException("fixture");
        }

        fixture.Customizations.Insert(
            0,
            new MockRelay(new IsDelegateSpecification()));
    }
}
```

The `IsDelegateSpecification` would then be defined as:

```csharp
public class IsDelegateSpecification : IRequestSpecification
{
    public bool IsSatisfiedBy(object request)
    {
        var t = request as Type;

        return t != null && IsDelegate(t);
    }

    private static bool IsDelegate(Type type)
    {
        return typeof(MulticastDelegate).IsAssignableFrom(type.BaseType);
    }
}
```

where `IsDelegate` is taken from the [internal extension method with the same name](https://github.com/AutoFixture/AutoFixture/blob/ab829640ed8e02776e4f4730d0e72ab3cc382339/Src/AutoMoq/MockType.cs#L46-L53).

It's also possible to combine `AutoMoqCustomization` and `EnableAutoMoqForDelegatesCustomization` in a _composite customization_ to make it easier to use:

```csharp
public class AutoMoqWithDelegatesCustomization : CompositeCustomization
{
    public AutoMoqWithDelegatesCustomization()
        : base(
            new AutoMoqCustomization(),
            new EnableAutoMoqForDelegatesCustomization())
    {
    }
}
```

## Demonstration

The final test becomes:

```csharp
[Fact]
public void FixtureCanFreezeMockOfFunc()
{
    var fixture = new Fixture().Customize(new AutoMoqWithDelegatesCustomization());
    var frozen = fixture.Freeze<Mock<Func<string>>>();

    var actual = fixture.Create<Func<string>>();
    
    Assert.Same(frozen.Object, actual); ✔️
}
```

which passes! 🎉

**Comment by zvirja** [comment]
Actually this is not a bug, but a lucking feature. It will be a huge breaking change to simply redirect all the delegate requests to Moq, as currently they _always_ generate the result value, which might not be the case for Moq.

In any case, I've created PR #999 to add this feature in the future versions of glue library, so you can enable it if you require that.

**Comment by zvirja** [comment]
Released in v4.2.0. To enable the feature:
```c#
fixture.Customize(new AutoMoqCustomization { GenerateDelegates = true });
```
